course5/homework10.2/linked_list.h

```
@@ -0,0 +1,148 @@
+		{
+			*current = std::make_unique<Node>();
+			(*current)->value = *it;
+			current = &((*current)->next);
+		}
+	}
+
+	LinkedList & operator=(LinkedList const & that)
+	{
+		this->head = nullptr;
+		std::unique_ptr<Node> * current{&head};
+		for (auto it = that.begin(); it != that.end(); ++it)
+		{
+			*current = std::make_unique<Node>();
+			(*current)->value = *it;
+			current = &((*current)->next);
+		}
+		return *this;
+	}
+		
+	LinkedList & operator=(LinkedList && that) { this->head = std::move(that.head); return *this; }
+
+	bool operator==(LinkedList const & that) const
+	{
+		auto itl = begin(), itr = that.begin();
+		for (; itl != end() && itr != that.end(); ++itl, ++itr)
+		{
+			if (*itl != *itr)
+				return false;
+		}
+		return itl == end() && itr == that.end();
+	}
+	bool operator!=(LinkedList const & that) const { return !(*this == that); }
+
+	LinkedList::iterator begin() { return iterator(head.get()); }
+	LinkedList::const_iterator begin() const { return const_iterator(head.get()); }
+	LinkedList::iterator end() { return iterator(nullptr); }
+	LinkedList::const_iterator end() const { return const_iterator(nullptr); }
+
+	void push_front(T const & element)
+	{
+		std::unique_ptr<LinkedList<T>::Node> newNode = std::make_unique<LinkedList<T>::Node>();
+		newNode->value = element;
+		newNode->next = std::move(head);
+		head = std::move(newNode);
+	}
+	void push_front(T && element)
+	{
+		std::unique_ptr<LinkedList<T>::Node> newNode = std::make_unique<LinkedList<T>::Node>();
+		newNode->value = std::move(element);
+		newNode->next = std::move(head);
+		head = std::move(newNode);
+	}
+
+	template <class... Args>
+	void emplace_front(Args&& ... args)
+	{
+		std::unique_ptr<LinkedList<T>::Node> newNode = std::make_unique<LinkedList<T>::Node>();
+		newNode->value = T(std::forward<Args>(args)...);
+		newNode->next = std::move(head);
+		head = std::move(newNode);
+	}
+
+	void pop_front() { if (head) head = std::move(head->next); }
```

The ```if``` breaks "don't pay for what you're not using" principle.

---

What's the best solution? Document that popping from an empty list is undefined and let the program crash?